packages/ember-data/lib/system/references/has-many.js

```diff
@@ -0,0 +1,84 @@
+import Ember from 'ember';
+import Reference from './reference';
+
+const get = Ember.get;
+
+var HasManyReference = function(store, parentInternalModel, hasManyRelationship) {
+  this._super$constructor(store, parentInternalModel);
+  this.hasManyRelationship = hasManyRelationship;
+  this.type = hasManyRelationship.relationshipMeta.type;
+
+  if (hasManyRelationship.link) {
+    this.remoteType = "link";
+    this.link = hasManyRelationship.link;
+  } else {
+    this.remoteType = "ids";
+    this._ids = hasManyRelationship.canonicalState.map(function(internalModel) {
+      return internalModel.id;
+    });
+  }
+
+  // TODO parent
+  // TODO inverse
+};
+
+HasManyReference.prototype = Ember.create(Reference.prototype);
+HasManyReference.prototype.constructor = HasManyReference;
+HasManyReference.prototype._super$constructor = Reference;
+
+HasManyReference.prototype.ids = function() {
+  return this._ids;
+};
+
+HasManyReference.prototype.meta = function() {
+  return this.hasManyRelationship.manyArray.meta;
+};
+
+HasManyReference.prototype.push = function(objectOrPromise) {
+  return Ember.RSVP.resolve(objectOrPromise).then((data) => {
+    var internalModels = data.map((payload) => {
+      var record = this.store.push(this.type, payload);
+      return record._internalModel;
+    });
+
+    this.hasManyRelationship.addCanonicalRecords(internalModels);
+
+    return this.hasManyRelationship.manyArray;
+  });
+};
+
+HasManyReference.prototype._isLoaded = function() {
+  var manyArray = this.hasManyRelationship.manyArray;
+
+  var hasData = get(this.hasManyRelationship, 'hasData');
+  var isEveryLoaded = manyArray.everyProperty('isLoaded', true);
+
+  return (hasData && isEveryLoaded);
+};
+
+HasManyReference.prototype.value = function() {
+  if (this._isLoaded()) {
+    return this.hasManyRelationship.manyArray;
+  }
+
+  return null;
+};
+
+HasManyReference.prototype.load = function() {
+  if (!this._isLoaded()) {
+    return this.hasManyRelationship.getRecords().get("promise");
+  }
+
+  var manyArray = this.hasManyRelationship.manyArray;
+  return Ember.RSVP.resolve(manyArray);
+};
+
+HasManyReference.prototype.reload = function() {
+  if (this._isLoaded()) {
+    return this.hasManyRelationship.manyArray.reload();
```

Comment:
One of the main reasons we are adding references is to differentiate between a manyArray that has records, [1,2,3] and the relationship, which we might want to reload from `/records/relationship` which returns record [3,4,5]